src/browser-connection/set.js

```diff
@@ -0,0 +1,178 @@
+    }
+
+    _cleanJobQueue (bc) {
+        if (bc instanceof LocalBrowserConnection)
+            bc.jobQueue = [];
+        else {
+            if (!bc.jobQueue.length || !this.task)
+                return;
+
+            if (this.task.pendingBrowserJobs.indexOf(bc.jobQueue[0]) > -1)
+                bc.removeJob(bc.jobQueue[0]);
+        }
+    }
+
+    _createConnectionIdlePromise (bc) {
+        return new Promise(resolve => {
+            if (bc.idle || bc.closed) {
+                resolve();
+                return;
+            }
+            var waitingForIdleTimeout = setTimeout(() => resolve(), this.WAITING_FOR_IDLE_TIMEOUT);
+
+            var eventHandler = () => {
+                clearTimeout(waitingForIdleTimeout);
+                resolve();
+            };
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
+        });
+    }
+
+    _createConnectionClosePromise (bc) {
+        return new Promise(resolve => {
+            if (bc.closed) {
+                resolve();
+                return;
+            }
+
+            bc.close();
+            bc.once('closed', resolve);
+        });
+    }
+
+    _createConnectionsReadyPromise () {
+        var ready = this.connections
+            .filter(bc => !bc.ready)
+            .map(bc => new Promise(resolve => bc.once('ready', resolve)));
+
+        return Promise.all(ready);
+    }
+
+    async _checkForDisconnectedConnections () {
+        var disconnectedUserAgents = this.connections
+            .filter(bc => bc.closed)
+            .map(bc => bc.userAgent);
+
+        if (disconnectedUserAgents.length) {
+            await this.closeConnections();
+
+            throw new Error(getText(MESSAGE.cantRunAgainstDisconnectedBrowsers, disconnectedUserAgents.join(', ')));
+        }
+    }
+
+    async _connectionErrorHandler (msg) {
+        await this.closeConnections();
+
+        this.emit('error', msg);
+    }
+
+    //API
+    async addConnections (...connections) {
+        connections      = flatten(connections);
+        this.connections = this.connections.concat(connections);
+
+        connections.forEach(bc => bc.once('error', this.bindedErrorHandler));
+
+        await this._checkForDisconnectedConnections();
+    }
+
+    closeConnection (bc) {
+        var promisedClose = null;
+
+        var asyncClose = async () => {
+            if (this.connections.indexOf(bc) === -1)
+                return;
+
+            bc.removeListener('error', this.bindedErrorHandler);
+
+            this._cleanJobQueue(bc);
+
+            if (bc.jobQueue.length)
+                return;
+
+            await this._createConnectionIdlePromise(bc);
+
+            if (bc instanceof LocalBrowserConnection)
+                await this._createConnectionClosePromise(bc);
+
+            remove(this.connections, bc);
+            remove(this.promisedCloses, promisedClose);
+        };
+
+        promisedClose = asyncClose(bc);
```

Comment:
I can't parse it. Why we have `promiseClose` which is equal to the `asyncClose`, please refactor it.